### ISSUE TITLE
Dramatic slowdown between versions

### ISSUE THREAD

**Issue by danielgordon10** [microsoft/pyright#1613]
Is there a reason that between version 1.1.117 and 1.1.118 the performance became significantly slower?

`npm install -g pyright@1.1.117`
`pyright  --stats`
```
21 errors, 0 warnings, 0 infos
Completed in 4.886sec

Analysis stats
Total files analyzed: 364

Timing stats
Find Source Files:    0sec
Read Source Files:    0.02sec
Tokenize:             0.26sec
Parse:                0.54sec
Resolve Imports:      1sec
Bind:                 0.64sec
Check:                2.24sec
Detect Cycles:        0.02sec
```

vs.

`npm install -g pyright@1.1.118`
`pyright  --stats`
```
21 errors, 0 warnings, 0 infos
Completed in 36.508sec

Analysis stats
Total files analyzed: 364

Timing stats
Find Source Files:    0sec
Read Source Files:    0.03sec
Tokenize:             0.26sec
Parse:                0.51sec
Resolve Imports:      1.25sec
Bind:                 0.66sec
Check:                33.64sec
Detect Cycles:        0.02sec
```

**Comment by erictraut** [comment]
This is due to a new feature called "implied else narrowing". I introduced a change in version 1.1.119 that should mitigate the perf impact to a large extent. Please rerun your measurements with the latest version (1.1.120) and let me know what you find.

**Comment by danielgordon10** [comment]
No difference
`pyright --version`
```
pyright 1.1.120
```

`pyright --stats`

```
...
21 errors, 0 warnings, 0 infos
Completed in 37.792sec

Analysis stats
Total files analyzed: 364

Timing stats
Find Source Files:    0.01sec
Read Source Files:    0.04sec
Tokenize:             0.26sec
Parse:                0.51sec
Resolve Imports:      1.42sec
Bind:                 0.68sec
Check:                34.73sec
Detect Cycles:        0.02sec


```

**Comment by erictraut** [comment]
Is your source code publicly available (e.g. in a github repo)? Or is it something you'd be willing to give me access to? If so, I can look at whether there are additional mitigations. If not, it's going to be difficult for me to take any further actions on this.

**Comment by erictraut** [comment]
I'm guessing that one or more of your source files has a very long chain of if/elif statements with no else at the end — and perhaps it's nested within one or more levels of for or while loops. If you can find a function or method that has those characteristics, perhaps you could extract it and create a self-contained sample.

**Comment by danielgordon10** [comment]
Unfortunately it is not publicly available, but thanks for the offer. I'll take a look for a giant if/else chain.

**Comment by danielgordon10** [comment]
I have found the largest source of slowdown

```
plots = List[Optional[np.ndarray]]
...
start_im = plots[col + cols * row]
if start_im is None:
    continue
im = start_im
...
many uses of im lower in the function
```
takes about 1 second to analyze

vs.
```
plots = List[Optional[np.ndarray]]
...
im = plots[col + cols * row]
if im is None:
    continue
...
many uses of im lower in the function
```
takes about 20 seconds to analyze

This is somewhat surprising to me but you may have better insight.

**Comment by erictraut** [comment]
Thanks for the additional information. That's a good clue, but it's not enough for me to fully understand the issue. Would you feel comfortable providing more of that function?

I presume that the code snippet in question is located within a loop, and probably a doubly-nested loop? I'm also guessing there are many dependencies between the local variables and that most of them do not contain type annotations, so the type analyzer is attempting to use code-flow analysis to infer their types.

Here's another experiment that would help us validate whether that particular `if` statement is at the heart of the problem. Please try adding an explicit `else` clause:

```python
if im is None:
    continue
else:
    pass
```

Does that change have an appreciable effect on the analysis time?

Thanks much for your help — I really appreciate it. I'm highly motivated to find the cause of the perf issue and mitigate it.

**Comment by danielgordon10** [comment]
Sure. Here is the code before the change.

```python
def subplot(
    plots: List[Optionalnp.ndarray]],
    rows: int,
    cols: int,
    output_width: Optional[int] = None,
    output_height: Optional[int] = None,
    border: int = DEFAULT_BORDER,
    titles: Optional[List[str]] = None,
    normalize: Optional[List[bool]] = None,
    order: Optional[List[int]] = None,
    border_color: Tuple[int, int, int] = (191, 191, 191),
) -> np.ndarray:
    """Given a list of images, returns a single image with the sub-images tiled in a subplot.

    Args:
        plots (List[np.ndarray]): List of images to plot. Can be of different sizes and dimensions as long as they are
            2 or 3 dimensional.
        rows (int): Number of rows in subplot. If there are fewer images than rows, it will add empty space for the
            blanks. If there are fewer rows than images, it will not draw the remaining images.
        cols (int): Number of columns in subplot. Similar to rows.
        output_width (Optional[int], optional): Width in pixels of a single subplot output image. If none is provided,
            the width of the first image is used. Defaults to None.
        output_height (Optional[int], optional): Height in pixels of a single subplot output image. If none is provided
            the height of the first image is used. Defaults to None.
        border (int, optional): Amount of border padding pixels between each image. Defaults to DEFAULT_BORDER.
        titles (Optional[List[str]], optional): Titles for each subplot to be rendered on top of images.
            Defaults to None.
        normalize (Optional[List[bool]], optional): List of whether to subtract the max and divide by the min before
            colorizing. If none, assumed false for all images. Defaults to None.
        order (Optional[List[int]], optional): If provided this reorders the provided plots before drawing them.
            Defaults to None.
        border_color (Tuple[int, int, int], optional): Defaults to (191, 191, 191).

    Returns:
        np.ndarray: A single image containing the provided images (up to rows * cols).
    """

    if order is not None:
        plots = [plots[im_ind] for im_ind in order]
        if titles is not None:
            titles = [titles[im_ind] for im_ind in order]
        if normalize is not None:
            normalize = [normalize[im_ind] for im_ind in order]

    output_width = plots[0].shape[1] if output_width is None else output_width
    output_height = plots[0].shape[0] if output_height is None else output_height
    assert output_width > 0
    assert output_height > 0

    returned_image = np.full(
        ((output_height + border) * rows + 2 * border, (output_width + border) * cols + 2 * border, 3),
        border_color,
        dtype=np.uint8,
    )
    for row in range(rows):
        for col in range(cols):
            try:
                if col + cols * row >= len(plots):
                    return returned_image
                im = plots[col + cols * row]
                if im is None:
                    continue
                im = im.squeeze()
                if im.dtype != np.uint8 or len(im.shape) < 3:
                    if normalize is None or normalize[col + cols * row]:
                        im = im.astype(np.float32)
                        im -= np.min(im)
                        im *= 255 / max(np.max(im), 1e-10)
                        im = im.astype(np.uint8)
                    else:
                        im = im.astype(np.uint8)
                if len(im.shape) < 3:
                    im = cv2.applyColorMap(im, cv2.COLORMAP_JET)[:, :, ::-1]

                im = min_side_resize_and_pad(im, output_height, output_width)
                if (
                    titles is not None
                    and len(titles) > 1
                    and len(titles) > col + cols * row
                    and len(titles[col + cols * row]) > 0
                ):
                    scale_factor = im.shape[1] / 320.0
                    if isinstance(titles[col + cols * row], str):
                        im = draw_contrast_text_cv2(
                            im, titles[col + cols * row], (30, int(30 * scale_factor)), 0.5 * scale_factor
                        )
                    else:
                        for tt, _ in enumerate(titles[col + cols * row]):
                            im = draw_contrast_text_cv2(
                                im,
                                titles[col + cols * row],
                                (30, int((tt + 1) * 30 * scale_factor)),
                                0.5 * scale_factor,
                            )
                returned_image[
                    border + (output_height + border) * row : (output_height + border) * (row + 1),
                    border + (output_width + border) * col : (output_width + border) * (col + 1),
                    :,
                ] = im
            except Exception as ex:
                print("Failed for image", col + cols * row)
                print("shape", plots[col + cols * row].shape)
                print("type", plots[col + cols * row].dtype)
                if titles is not None and len(titles) > col + col * row:
                    print("title", titles[col + cols * row])
                raise ex

    im = returned_image
    # for one long title
    if titles is not None and len(titles) == 1:
        scale_factor = max(max(im.shape[0], im.shape[1]) * 1.0 / 600, 1)
        cv2.putText(im, titles[0], (10, 30), CV_FONT, 0.5 * scale_factor, [0, 0, 0], 4)
        cv2.putText(im, titles[0], (10, 30), CV_FONT, 0.5 * scale_factor, [255, 255, 255], 1)

    return im
```

**Comment by erictraut** [comment]
I'm able to repro the slow-down, and I'll work on diagnosing the problem. Thanks much for your help!

**Comment by erictraut** [comment]
The type analyzer was recalculating some information many times that could be safely cached. With a simple cache in place, the analysis time for the above code dropped from 17100ms to just over 200ms on my machine. This performance fix will be included in the next release of pyright and pylance. Thanks again for all of your help!

**Comment by danielgordon10** [comment]
Cool, glad I could help. Thanks for being so receptive.

**Comment by erictraut** [comment]
I'm going to leave this issue open until the next release.

**Comment by erictraut** [comment]
This change is included in pyright 1.1.122, which I just published. It will also be in the next release of pylance.
